File: Tests/iOS Tests/Tests/PNSubscribeToPresenceChannelsTests.m
//
//  PNSubscribeToPresenceChannelsTests.m
//  PubNub Tests
//
//  Created by Jordan Zucker on 6/30/15.
//
//

#import <PubNub/PubNub.h>

#import "PNBasicSubscribeTestCase.h"

@interface PNSubscribeToPresenceChannelsTests : PNBasicSubscribeTestCase

@end

@implementation PNSubscribeToPresenceChannelsTests

- (BOOL)isRecording{
    return NO;
}

- (NSArray *)subscriptionChannels {
    return @[
             @"a"
             ];
}

- (void)tearDown {
    PNWeakify(self);
    self.didReceiveStatusAssertions = ^void (PubNub *client, PNSubscribeStatus *status) {
        PNStrongify(self);
        XCTAssertEqualObjects(self.client, client);
        XCTAssertNotNil(status);
        XCTAssertFalse(status.isError);
        //        XCTAssertEqual(status.operation, PNUnsubscribeOperation);
        //        XCTAssertEqual(status.category, PNDisconnectedCategory);
        //        XCTAssertEqual(status.subscribedChannels.count, 0);
        XCTAssertEqual(status.subscribedChannelGroups.count, 0);
        XCTAssertEqual(status.operation, PNUnsubscribeOperation);
        NSLog(@"timeToken: %@", status.currentTimetoken);
        //        XCTAssertEqualObjects(status.currentTimetoken, @14355626738514132);
        //        XCTAssertEqualObjects(status.currentTimetoken, status.data.timetoken);
        [self.unsubscribeExpectation fulfill];
        
    };
    [self PNTest_unsubscribeFromChannels:[self subscriptionChannels] withPresence:YES];
    [super tearDown];
}

- (void)testSimpleSubscribeToPresenceChannels {
    PNWeakify(self);
    self.didReceiveStatusAssertions = ^void (PubNub *client, PNSubscribeStatus *status) {
        PNStrongify(self);
        XCTAssertEqualObjects(self.client, client);
        XCTAssertNotNil(status);
        XCTAssertFalse(status.isError);
        XCTAssertEqual(status.category, PNConnectedCategory);
        XCTAssertEqual(status.subscribedChannelGroups.count, 0);
        NSArray *expectedPresenceSubscriptions = @[@"a-pnpres"];
        XCTAssertEqualObjects([NSSet setWithArray:status.subscribedChannels],
                              [NSSet setWithArray:expectedPresenceSubscriptions]);
        XCTAssertEqual(status.operation, PNSubscribeOperation);
        NSLog(@"timeToken: %@", status.currentTimetoken);
        XCTAssertEqualObjects(status.currentTimetoken, @14508182981759319);
        XCTAssertEqualObjects(status.currentTimetoken, status.data.timetoken);
        [self.subscribeExpectation fulfill];
    };
    [self PNTest_subscribeToPresenceChannels:[self subscriptionChannels]];
}

@end

File: Tests/iOS Tests-Swift/PNSubscribeToPresenceChannelsTests.swift
//
//  PNSubscribeToPresenceChannelsTests.swift
//  PubNub Tests
//
//  Created by Vadim Osovets on 10/20/15.
//
//

import UIKit
import XCTest

import Foundation

class PNSubscribeToPresenceChannelsTests: PNBasicSubscribeTestCase {
    
    let subscriptionChannels = ["a"]
    
    override func isRecording() -> Bool {
        return false
    }
    
    override func tearDown() {
        
        self.didReceiveStatusAssertions = {(client: PubNub, status: PNStatus) -> (Void) in
            
            XCTAssertNotNil(status)
            XCTAssertEqual(self.client, client, "Incorrect client")
            XCTAssertFalse(status.error)
            
            let subscribedStatus = status as? PNSubscribeStatus
            
            if (subscribedStatus != nil) {
                XCTAssertEqual(subscribedStatus?.subscribedChannelGroups.count, 0)
            } else {
                // TODO: investigate me
                //                XCTFail("Unexpected status")
            }
            
            //            XCTAssertTrue(status.category == .UnsubscribeOperation)
            
            self.unsubscribeExpectation.fulfill()
        }
        
        self.PNTest_unsubscribeFromChannels(subscriptionChannels, presense: true)
        
        super.tearDown()
    }
    
    func testSimpleSubscribeToPresenceChannels() {
        
        self.didReceiveStatusAssertions = {(client: PubNub, status: PNStatus) -> (Void) in
            XCTAssertNotNil(status)
            XCTAssertEqual(self.client, client, "Incorrect client")
            XCTAssertFalse(status.error)
            XCTAssertTrue(status.operation == .SubscribeOperation)
            
            let expectedChannels: [String] = ["a-pnpres"]
            
            let subscribedStatus = status as? PNSubscribeStatus
            
            if (subscribedStatus != nil) {
                var resultSet: Set<String> = []
                for element in (subscribedStatus?.subscribedChannels as? [String])! {
                    resultSet.insert(element)
                }
                
                let expectedChannelsSet: Set<String> = Set(expectedChannels)
                
                XCTAssertTrue(resultSet == expectedChannelsSet, "Subscribed channel groups list are not equal")
                
                XCTAssertEqual(subscribedStatus?.currentTimetoken, subscribedStatus?.data.timetoken)
                
                XCTAssertEqual(subscribedStatus?.currentTimetoken, NSDecimalNumber(string : "14356558880455349"))
                XCTAssertEqual(subscribedStatus?.currentTimetoken, subscribedStatus?.data.timetoken)
            }
            
            XCTAssertTrue(status.operation == .SubscribeOperation)
            
            self.subscribeExpectation?.fulfill()
        }
        
        self.PNTest_subscribeToPresenceChannels(self.subscriptionChannels)
    }
}
